hack/upgrade-cnv.sh

```diff
@@ -1,3 +1,105 @@
+
+echo "creating subscription"
+oc create -f - <<EOF
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+metadata:
+  name: kubevirt-hyperconverged
+  namespace: $TARGET_NAMESPACE
+  labels:
+    operators.coreos.com/kubevirt-hyperconverged.openshift-cnv: ''
+spec:
+  channel: stable
+  installPlanApproval: Automatic
+  name: kubevirt-hyperconverged
+  source: redhat-operators
+  sourceNamespace: openshift-marketplace
+EOF
+
+echo "waiting for HyperConverged operator to become ready"
+$SCRIPT_DIR/wait-for-hco.sh
+
+#=======================================
+# Upgrade HCO to latest build from brew
+#=======================================
+
+OLD_CSV=$(oc get subscription kubevirt-hyperconverged -n $TARGET_NAMESPACE -o jsonpath="{.status.installedCSV}")
+
+echo "waiting for the previous CSV installation to complete"
+$SCRIPT_DIR/retry.sh 60 10 "oc get ClusterServiceVersion $OLD_CSV -n $TARGET_NAMESPACE -o jsonpath='{.status.phase}' | grep 'Succeeded'"
+
+echo "setting up brew catalog source"
+$SCRIPT_DIR/create-brew-catalogsource.sh
+
+echo "patching the subscription to switch to the brew catalog source"
+oc patch subscription kubevirt-hyperconverged -n $TARGET_NAMESPACE --type=merge --patch='{"spec": {"source": "brew-catalog-source"}}'
+
+echo "waiting for the new CSV installation to commence"
+sleep 60
+
+echo "waiting for the subscription's currentCSV to move to the new catalog source"
+$SCRIPT_DIR/retry.sh 30 10 "oc get subscription kubevirt-hyperconverged -n $TARGET_NAMESPACE -o jsonpath='{.status.currentCSV}' | grep -v $OLD_CSV"
+
+NEW_CSV=$(oc get subscription kubevirt-hyperconverged -n $TARGET_NAMESPACE -o jsonpath='{.status.currentCSV}')
+
+echo "waiting for HyperConverged operator to become ready"
+$SCRIPT_DIR/wait-for-hco.sh
+
+echo "waiting for the subscription's installedCSV to move to the new catalog source"
+$SCRIPT_DIR/retry.sh 60 10 "oc get subscription kubevirt-hyperconverged -n $TARGET_NAMESPACE -o jsonpath='{.status.currentCSV}' | grep $NEW_CSV"
+
+# oc currently fails if the resource doesn't exist while the command is executed;
+# this will be resolved in kubectl v1.21: https://github.com/kubernetes/kubernetes/pull/96702
+# till then, we manually test for 404 in the output string (both with v1.20 and pre-v.120 formats)
+echo "waiting for the previous CSV to be completely removed"
+WAIT_CSV_OUTPUT=$(oc wait ClusterServiceVersion $OLD_CSV -n $TARGET_NAMESPACE --for delete --timeout=10m 2>&1)
+if [ $? -ne 0 ] && ! grep -qE "NotFound|(no matching resources found)" <(echo "$WAIT_CSV_OUTPUT"); then
+  echo "$WAIT_CSV_OUTPUT"
+  exit 1
+fi
+echo "$WAIT_CSV_OUTPUT"
+
+# The previous CSV get deleted with "background deletion" propagation
+# policy, i.e. before its owned resources are completely removed.
+# Therefore, we wait for a few more minutes to let these resources be
+# completely deleted, to avoid any conflicts in the subsequent test execution.
+echo "waiting for residual resources to complete deletion"
+sleep 10m
+
```

Comment:
Shouldn't we at least check whether everything is gone that should not exist any more? Is that not possible?

Reply:
It's possible, but quite complex with bash (see operator-framework/operator-lifecycle-manager#2007).
Unfortunately, there's no identifying label on the old CSV pods associating them with the old CSV.
Hence, we'll need to fetch all deployments owned by the old CSV, then fetch each of the deployments' owned replicasets, and then fetch each of these replicasets' pods. Then verify each of these pods is gone.